### ISSUE TITLE
Slippage and Commission Units?

### ISSUE THREAD

**Issue by Ericxgao** [enigmampc/catalyst#145]
There isn't too much documentation on this, but I'd like clarification on the units for slippage and commission.

    context.set_commission(maker=.001, taker=.001)
    context.set_slippage(spread=0.0000001)

For example - does the above set a 0.1% commission on both maker and taker orders?

When calculating slippage, is it creating a 10 satoshi price difference from the provided price on the candlestick?

**Comment by caioaao** [comment]
Be extremely careful with that. Setting a commission on my algorithm made it go from a bad algorithm to one that returns 200x on backtest. Lost some money testing it live

**Comment by fredfortier** [comment]
@caioaao thank you for reporting this and sorry about this issue. I'm investigating now.

**Comment by fredfortier** [comment]
I have not identified this issue yet: 
> Setting a commission on my algorithm made it go from a bad algorithm to one that returns 200x on backtest

I'm sharing this update to establish a baseline. I'm trying to recreate this issue in an algo using: `mean_reversion_simple.py`. I've set my commission like this: `context.set_commission(maker=0.001, taker=0.002)`. This is the same as the default commissions for the `eth_btc` market used in this backtest.

I took the first transaction (buy order) and broke down how the algo calculates the cost basis and validate each calculation individually.
```
Date: 2017-10-01 00:00:00+00:00 
Price: 0.06951499
Limit price: 0.06986256495
Amount: Capital (0.1) / Price (0.06951499) = 1.43853865188

Date next bar: 2017-10-01 00:01:00+00:00
Next bar price: 0.069547
Next bar under limit: Yes

Spread: 0.0001
Execution price: 0.069547 * 1.0001 = 0.0695539547

Cost: Amount (1.43853865188) * Price (0.0695539547) = 0.100056052227
Muptiplier: Taker (0.002), the order executed below limit price so it does not get a maker discount
Fee: Cost (0.100056052227) * Multiplier (0.002) = 0.000200112113959
Cost Basis: Execution price (0.0695539547) + Fee (0.000200112113959) = 0.06969306
```
In this one example, the calculations appear accurate as demonstrated here. It arrives to the exact same result whether I set the commission or use the default. I must be missing the condition causing error. I will continue to look for it and add these validation steps to unit tests to ensure that this issue never occurs in the future.

**Comment by fredfortier** [comment]
@caioaao I'm still trying to find the error. I can confirm that commissions get calculated correctly whenever Zipline calls: `commission.calculate(order, txn)` including here: `catalyst/finance/blotter.py:371`. When I pull the list of transactions in `analyze()`, I do see that the commission attribute is None. However, I can't see where Zipline applies the commission to the transaction objects, like it does to orders in the method referenced above. I continue to investigate.

**Comment by caioaao** [comment]
I think I found out the issue, at least with poloniex. In Poloniex you always have to set the limit price, but it doesn't mean it's a limit order. If the price is contained in the order book (eg: buying above market value or selling below market value), it's considered a market taking order. When setting the commissions with `set_commission(maker=0.0015, taker=0.0025)`, the backtester will always choose the market maker commission as all orders must have a limit price, thus improving the backtest results. Setting both to `0.0025` does the trick as it will be the worst case scenario for an algorithm. Probably leaving this as the default for poloniex is a good idea.

**Comment by fredfortier** [comment]
Yes, that's how other exchanges work as well. When you make a limit order, you potentially add liquidity to the order book, but you are only rewarded if someone takes your order. If you don't reach the limit price and the order still closes, it implicitly means that you took someone else's order and your limit order is deleted from the order book.

Unless there is an issue which I could not find, the simulation accounts for this. Specifically, it checks the open order triggers before each frame. In the frame following your order, if the price is lower than your limit, it will close your order and apply the taker fee. If the price reaches your limit, it applies the maker fee. You are correct that this is somewhat biased, it reality someone else may have taken these orders first leaving yours open. Or, the price may have briefly dipped below the limit price, making you pay the taker fees.

A more precise simulation model could evaluate all the order book entries between the frames. This would involve a huge amount of data, difficult to impossible to collect with current exchange APIs, as it would need to retrace all changes to order book and not sure the executed orders. This would still be biased because we can't precisely model the past, the algo would need to assume that you were able to take a particular order before whoever took it in reality. This assumption would be valid only some of the time.

I apologize if I'm stating something which you already know, but I think that this discussion is helpful. We'll write an article about this in our documentation. I'll verify again with a focus on this particular behavior to identify any error.

We appreciate your time looking into this.

**Comment by lacabra** [comment]
@fredfortier reminder to add this to the documentation as per your comment above

**Comment by eranhirs** [comment]
I am using Binance backtesting, what is the default maker / taker fees catalyst is using?

Debugging `context.blotter.commission_models` & `context.blotter.slippage_models`, it seems that slippage is by default 0.0001, and maker_fee and taker_fee are set to None.
